Azd updates (#468)
dotnet/docs-aspire: docs/whats-new/preview-4.md

## Azure Developer CLI

The [Azure Developer CLI](https://aka.ms/azd) (azd) has been updated to support the new manifest types introduced in preview 4. Azd supports prompting for parameter resources, and supports the bicep resource which enables deploying any thing. All Azure resources can be described using a bicep module.

In addition to new features, azd will automatically create secrets in Azure Container Apps for any parameter marked as a secret, and any environment variable that references a secret or uses a connection string.

> ### Prompt for parameter

Reviewer: Prompt, or show how the environment variable can be set.

When azd detects a parameter, it will prompt for a value. For example, having the next AppHost code:

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var administratorLogin = builder.AddParameter("administratorLogin");
var administratorLoginPassword = builder.AddParameter("administratorLoginPassword", secret: true);
var pg = builder.AddPostgres("postgres")
                .AsAzurePostgresFlexibleServer(administratorLogin, administratorLoginPassword)
                .AddDatabase("db");

builder.AddProject<Projects.BicepSample_ApiService>("api")
       .WithReference(pg);
```

Two parameters are defined to set the Postgres flexible server administrator login. Then, azd will ask for both parameters.

> :::image type="content" source="media/preview-4/promptUser.gif" lightbox="media/preview-4/promptUser.gif" alt-text="azd parameters prompting":::

Reviewer: Show a screenshot of how this gets stored on the ACA app

## Azure Developer CLI

The [Azure Developer CLI](https://aka.ms/azd) (azd) has been updated to support the new manifest types introduced in preview 4. Azd supports prompting for parameter resources, and supports the bicep resource which enables deploying any thing. All Azure resources can be described using a bicep module.

In addition to new features, azd will automatically create secrets in Azure Container Apps for any parameter marked as a secret, and any environment variable that references a secret or uses a connection string.

### Prompt for parameter

When azd detects a parameter, it will prompt for a value. For example, having the next AppHost code:

```csharp

> var builder = DistributedApplication.CreateBuilder(args);
> 
> var administratorLogin = builder.AddParameter("administratorLogin");
> var administratorLoginPassword = builder.AddParameter("administratorLoginPassword", secret: true);
> var pg = builder.AddPostgres("postgres")
>                 .AsAzurePostgresFlexibleServer(administratorLogin, administratorLoginPassword)
>                 .AddDatabase("db");
> 
> builder.AddProject<Projects.BicepSample_ApiService>("api")
>        .WithReference(pg);

Reviewer 2: Format code to align with the updates in the [manifest article](https://github.com/dotnet/docs-aspire/pull/440/files#diff-29685b2d8a0635d9a8215fe2d04cd41a7df7e73c9b913e239e58c8781bb3cc72R858-R867):

```suggestion
var builder = DistributedApplication.CreateBuilder(args);

var administratorLogin = builder.AddParameter("administratorLogin");
var administratorLoginPassword = builder.AddParameter(
    "administratorLoginPassword", secret: true);

var pg = builder.AddPostgres("postgres")
                .AsAzurePostgresFlexibleServer(
                    administratorLogin, administratorLoginPassword)
                .AddDatabase("db");
```

```json
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "administratorLogin": {
      "value": "${AZURE_ADMINISTRATOR_LOGIN}"
    },
    "administratorLoginPassword": {
      "value": "${AZURE_ADMINISTRATOR_LOGIN_PASSWORD}"
    }
  }
}
```

### Finding parameters after provision

Azd sets the parameter values for a project as part of the [App Container configuration](https://learn.microsoft.com/azure/container-apps/containers#configuration), using its environment. When the parameter is defined as `secret: true` (like `administratorLoginPassword` in the previous example), azd creates a secret for the App Container configuration and then links the environment to that secret:

:::image type="content" source="media/preview-4/secretValue.png" lightbox="media/preview-4/secretValue.png" alt-text="azd secret parameter":::

> On the previous example, azd takes the bicep resource that was generated by Aspire and set the input parameters for it:

Reviewer 2: ```suggestion
In the previous example, azd takes the bicep resource that was generated by .NET Aspire and sets the input parameter value for it:
```

```bicep
module postgres 'postgres2/aspire.hosting.azure.bicep.postgres.bicep' = {
  name: 'postgres'
  scope: rg
  params: {
    location: location
    administratorLogin: administratorLogin
    administratorLoginPassword: administratorLoginPassword
    databases: ['db']
    keyVaultName: resources.outputs.SERVICE_BINDING_POSTGRES2KV_NAME
    serverName: 'postgres'
  }
}
```

> Note how `azd` automatically generates a Key Vault for the postgres module to securely persist the connection string using a secret. When the project is deployed to a container app, the secret is used to set the full connection string as a container app secret (see picture above).

Reviewer 2: ```suggestion
In this case, `azd` automatically generates an Azure Key Vault resource for the Postgres module to securely persist the connection string using a secret. When the project is deployed to Container Apps, the secret is used to set the full connection string as a Container App secret as shown in the following image.
```

> If the secret is referenced directly, the container app secret will be a reference to the Key Vault Secret. See below.

Reviewer 2: ```suggestion
If the secret is referenced directly, the Container App secret will be a reference to the Key Vault secret as shown in the following image:
```

:::image type="content" source="media/preview-4/secretAndRefToSecret.png" lightbox="media/preview-4/secretAndRefToSecret.png" alt-text="azd secret and secret reference configuration":::

> Note that `connectionstring--server` holds the reference for the Key Vault secret. The secret is pulled by container app when it is used by the service project. On the other hand, the `connectionstring--db2` was created pulling the `connectionString` secret from Key Vault and adding the `Database=` configuration.

Reviewer 2: ```suggestion
The `connectionstring--server` holds the reference for the Key Vault secret. The secret is pulled by the Container App when it's used by the service project. On the other hand, the `connectionstring--db2` was created pulling the `connectionString` secret from Key Vault and adding the `Database=` configuration.
```

## Azure Developer CLI

The [Azure Developer CLI](https://aka.ms/azd) (azd) has been updated to support the new manifest types introduced in preview 4. Azd supports prompting for parameter resources, and supports the bicep resource which enables deploying any thing. All Azure resources can be described using a bicep module.

In addition to new features, azd will automatically create secrets in Azure Container Apps for any parameter marked as a secret, and any environment variable that references a secret or uses a connection string.

> ### Prompt for parameter

Reviewer: ```suggestion
### Prompting for parameters
```

> When azd detects a parameter, it prompts for a value. For example, consider the following app host code:

Reviewer: ```suggestion
When azd detects a parameter.v0 resource in the manifest, it will prompt for a value. For example, consider the following code in the apphost project:
```

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var administratorLogin = builder.AddParameter("administratorLogin");
var administratorLoginPassword = builder.AddParameter(
    "administratorLoginPassword", secret: true);

var pg = builder.AddPostgres("postgres")
                .AsAzurePostgresFlexibleServer(
                    administratorLogin, administratorLoginPassword)
                .AddDatabase("db");
```

In the preceding code, two parameters are created to represent the admin login and password. They're then passed to the `AsAzurePostgresFlexibleServer` API. This causes `azd` to ask for both parameter values.

:::image type="content" source="media/preview-4/promptUser.gif" lightbox="media/preview-4/promptUser.gif" alt-text="azd parameters prompting":::

> To skip interactive prompting, set an environment variable for each parameter. For the `administratorLogin` parameter, azd looks for an environment variable with the name `AZURE_ADMINISTRATOR_LOGIN`.

Reviewer: ```suggestion
To skip interactive prompting, users can set an environment variable for each parameter. For the `administratorLogin` parameter, azd looks for an environment variable with the name `AZURE_ADMINISTRATOR_LOGIN`.
```

> To skip interactive prompting, users can set an environment variable for each parameter. For the `administratorLogin` parameter, azd looks for an environment variable with the name `AZURE_ADMINISTRATOR_LOGIN`.

Reviewer: Does AZD magically turn the camel cased `administratorLogin` into `AZURE_ADMINISTRATOR_LOGIN`? What is the rule for this.

Author: Yes, azd makes it automatically.
Folks can set any name for the env var, there's no rule.

> [!TIP]
> These environment variables should be prefixed with `AZURE_` and followed by the name of the parameter, converted from camel case to upper-snake case.

To see the parameter mapping, run `azd infra synth` and view the `main.parameters.json` file in the _/infra_ folder.

```json
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "administratorLogin": {
      "value": "${AZURE_ADMINISTRATOR_LOGIN}"
    },
    "administratorLoginPassword": {
      "value": "${AZURE_ADMINISTRATOR_LOGIN_PASSWORD}"
    }
  }
}
```

> ### Find parameters after provision

Reviewer: ```suggestion
### Find parameters values after provisioning
```
